### PR TITLE
Move querystring tests from tide

### PR DESCRIPTION
**Pull request by plippe**
https://github.com/http-rs/http-types/issues/154

### CODE REVIEW

File: tests/querystring.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    assert_eq!(params.unwrap().msg, "Hello");
+}
+
+#[test]
+fn unsuccessfully_deserialize_query() {
+    let req = http_types::Request::new(Method::Get, Url::parse("http://example.com/").unwrap());
+
+    let params = req.query::<Params>();
+    assert!(params.is_err());
+    assert_eq!(params.err().unwrap().to_string(), "invalid data");
```
Comment (plippe) [inline tests/querystring.rs:33]: Different error message to the one in [original test](https://github.com/http-rs/tide/blob/master/tests/querystring.rs#L64)

File: tests/querystring.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        Url::parse("http://example.com/?error=should_fail").unwrap(),
+    );
+
+    let params = req.query::<Params>();
+    assert!(params.is_err());
+    assert_eq!(params.err().unwrap().to_string(), "missing field `msg`");
+}
+
+#[test]
+fn empty_query_string_for_struct_with_no_required_fields() {
```
Comment (plippe) [inline tests/querystring.rs:49]: Test fails

**Comment by plippe** [comment]
The failing test and the different error message are related to the missing query string. The tests are the same once I add any query string to both: e.g. `http://example.com/?`.
This is due to the query function that [fails when no query is present](https://github.com/http-rs/http-types/blob/master/src/request.rs#L634). 

I am unsure of what would be preferred:
- Should I add a query string to all tests?
- Should the query function return an `Option<Result<T>>`?

**Comment by plippe** [comment]
I also believe that the `query` function might be worth moving to the `Url` struct instead of the `Request`.

**Comment by plippe** [comment]
Completely missed that "default to an empty string".
It should work now (famous last words).

Also wondering about moving the implementation to the `Url`. It feels wrong to need a request to extract a URL's query string.

**Comment by yoshuawuyts** [comment]
> Also wondering about moving the implementation to the Url

It could be worth bringing up. However `url` ships with a decoder already, just not the `serde` bounds. It may just be they don't want the dependency, even if optional.

---

That said this patch looks great; let's merge it!

**Pull request merged**